File: strands-ts/src/models/usage.ts
/**
 * Normalization of provider-reported token counts into the SDK's {@link Usage} contract.
 *
 * Model providers report prompt-cache tokens under two incompatible conventions:
 *
 * - **Disjoint** (Bedrock Converse, the Anthropic API): the input field carries only the tokens
 *   that were neither read from nor written to the cache, and the cache counters are additional
 *   tokens. The full prompt is `input + cacheRead + cacheWrite`.
 * - **Inclusive** (OpenAI Chat Completions and Responses, Google, LiteLLM): the input field
 *   carries the *whole* prompt and the cache counters break out a subset of it. The full prompt
 *   is just `input`.
 *
 * {@link Usage} is defined as disjoint, so an inclusive provider's counts must have their cache
 * tokens subtracted out. Routing every adapter through {@link normalizeUsage} keeps the invariant
 * in one place instead of re-deriving it per adapter.
 *
 * @internal
 */

import { logger } from '../logging/logger.js'
import type { Usage } from './streaming.js'

/**
 * Provider-reported token counts, before normalization.
 *
 * @internal
 */
export interface RawUsage {
  /** The provider's prompt token count. */
  inputTokens: number | undefined
  /** The provider's completion token count, including reasoning tokens. */
  outputTokens: number | undefined
  /** Prompt tokens served from the cache. */
  cacheReadTokens?: number | undefined
  /** Prompt tokens written to the cache. */
  cacheWriteTokens?: number | undefined
  /** Tokens spent on internal reasoning; a subset of `outputTokens`. */
  reasoningTokens?: number | undefined
  /**
   * Whether `inputTokens` already contains the cache counts. Pass `true` for providers reporting
   * the full prompt in their input field (OpenAI, Google, LiteLLM), `false` for providers
   * reporting only net new tokens (Bedrock Converse, Anthropic).
   */
  inputIncludesCache: boolean
}

/**
 * A number in the decimal forms JSON can carry, which is the overlap of both SDKs' parsing.
 *
 * Every character is spelled out rather than left to `\d` and `trim()`, which do not accept the
 * same set as their Python counterparts.
 *
 * The fraction nests inside the integer alternative rather than following it, so a long run of
 * digits with an invalid tail is retried once per digit rather than once per split point, which
 * would cost time quadratic in the length of a count a gateway could send.
 */
const DECIMAL_NUMBER = /^[ \t\n\r\f\v]*[+-]?(?:[0-9]+(?:\.[0-9]*)?|\.[0-9]+)(?:[eE][+-]?[0-9]+)?[ \t\n\r\f\v]*$/

/**
 * Coerces a provider-reported token count to a non-negative integer.
 *
 * Providers omit counters they do not populate, and OpenAI-compatible gateways occasionally
 * return a string. Anything non-numeric is treated as absent so a malformed count degrades the
 * metric instead of throwing mid-stream.
 *
 * @param value - The raw value reported by the provider
 * @returns The value as a non-negative integer, or `0` when absent or not numeric
 *
 * @internal
 */
export function asTokenCount(value: unknown): number {
  let parsed = value
  if (typeof parsed === 'string') {
    // Matching first keeps the two SDKs agreeing on every string: Number() would otherwise take
    // the "0x10" prefixes Python's float() rejects, and float() would take the "1_000" separators
    // Number() rejects.
    if (!DECIMAL_NUMBER.test(parsed)) {
      return 0
    }
    parsed = Number(parsed)
  }
  // Past what JavaScript represents exactly a count is read as absent rather than silently
  // rounded, which is also where the Python SDK stops so the two report the same number.
  if (typeof parsed !== 'number' || !Number.isFinite(parsed) || parsed > Number.MAX_SAFE_INTEGER) {
    return 0
  }
  return Math.max(0, Math.trunc(parsed))
}

/**
 * Reads a cache hit off an OpenAI-compatible usage payload, whichever name carries it.
 *
 * A gateway fronting Anthropic lifts the two cache counters onto the usage object as a pair, and
 * builds the details payload only from counts that arrive as integers — so one reporting a count as
 * a JSON float leaves the details absent while both counters sit on the usage object. A
 * DeepSeek-shaped gateway spells the same count `prompt_cache_hit_tokens`.
 *
 * Reading a single name leaves a hit folded inside the prompt count, billed at the full input rate
 * rather than the cache-read rate, and nothing downstream can see it: the counters still sum to the
 * reported total, so the disjointness check does not fire.
 *
 * @param usage - The vendor usage payload, which may carry the count as a property
 * @param promptDetails - The payload's prompt token-details breakdown
 * @returns The number of cache read tokens, or `0` when the payload reports none
 *
 * @internal
 */
export function cacheReadCount(usage: unknown, promptDetails: unknown): number {
  const onUsage = usage as { cache_read_input_tokens?: unknown; prompt_cache_hit_tokens?: unknown } | null | undefined
  return (
    asTokenCount((promptDetails as { cached_tokens?: unknown } | null | undefined)?.cached_tokens) ||
    asTokenCount(onUsage?.cache_read_input_tokens) ||
    asTokenCount(onUsage?.prompt_cache_hit_tokens)
  )
}

/**
 * Reads a cache write off an OpenAI-compatible usage payload, whichever name carries it.
 *
 * The same count arrives under three names. A gateway fronting Anthropic lifts it to the usage
 * object as `cache_creation_input_tokens`, while one speaking the OpenAI shape leaves it on the
 * prompt details under either its own `cache_creation_tokens` or OpenAI's `cache_write_tokens`.
 *
 * More than one can carry a value at once, so the order is the contract rather than a convenience:
 * the count on the usage object is read first because it is the one the gateway itself derived, and
 * a details payload can carry a differently-derived number under the same name.
 *
 * Reading a single name leaves a write folded inside the prompt count, billed at the full input
 * rate rather than the cache-write rate, and nothing downstream can see it: the counters still sum
 * to the reported total, so the disjointness check does not fire.
 *
 * @param usage - The vendor usage payload, which may carry the count as a property
 * @param promptDetails - The payload's prompt token-details breakdown
 * @returns The number of cache write tokens, or `0` when the payload reports none
 *
 * @internal
 */
export function cacheWriteCount(usage: unknown, promptDetails: unknown): number {
  const onUsage = (usage as { cache_creation_input_tokens?: unknown } | null | undefined)?.cache_creation_input_tokens
  const details = promptDetails as { cache_creation_tokens?: unknown; cache_write_tokens?: unknown } | null | undefined
  return (
    asTokenCount(onUsage) || asTokenCount(details?.cache_creation_tokens) || asTokenCount(details?.cache_write_tokens)
  )
}

/**
 * Normalizes provider token counts into a disjoint {@link Usage}.
 *
 * The cache and reasoning fields are set only when non-zero, so a provider that does not report
 * them produces the same shape it always has.
 *
 * @param raw - The provider-reported counts and the convention they follow
 * @returns A `Usage` whose four billed counters are disjoint and sum to `totalTokens`
 *
 * @internal
 */
export function normalizeUsage(raw: RawUsage): Usage {
  const inputTokens = asTokenCount(raw.inputTokens)
  const outputTokens = asTokenCount(raw.outputTokens)
  const cacheReadTokens = asTokenCount(raw.cacheReadTokens)
  const cacheWriteTokens = asTokenCount(raw.cacheWriteTokens)
  const reasoningTokens = asTokenCount(raw.reasoningTokens)

  let netInputTokens = inputTokens
  if (raw.inputIncludesCache) {
    netInputTokens = inputTokens - cacheReadTokens - cacheWriteTokens
    if (netInputTokens < 0) {
      // A provider reporting cache counts exceeding its own prompt total is self-inconsistent;
      // clamp so the invariant holds rather than emit a negative.
      logger.warn(
        `input_tokens=<${inputTokens}>, cache_read_tokens=<${cacheReadTokens}>, ` +
          `cache_write_tokens=<${cacheWriteTokens}> | cache tokens exceed reported input tokens ` +
          `| clamping net input to zero`
      )
      netInputTokens = 0
    }
  }

  return {
    inputTokens: netInputTokens,
    outputTokens,
    totalTokens: netInputTokens + outputTokens + cacheReadTokens + cacheWriteTokens,
    ...(cacheReadTokens > 0 && { cacheReadInputTokens: cacheReadTokens }),
    ...(cacheWriteTokens > 0 && { cacheWriteInputTokens: cacheWriteTokens }),
    ...(reasoningTokens > 0 && { reasoningOutputTokens: reasoningTokens }),
  }
}

File: strands-py/src/strands/models/_usage.py
"""Normalization of provider-reported token counts into the SDK's :class:`Usage` contract.

Model providers report prompt-cache tokens under two incompatible conventions:

- **Disjoint** (Bedrock Converse, the Anthropic API): the input field carries only the tokens
  that were neither read from nor written to the cache, and the cache counters are additional
  tokens. The full prompt is ``input + cacheRead + cacheWrite``.
- **Inclusive** (OpenAI Chat Completions and Responses, Google, LiteLLM): the input field
  carries the *whole* prompt and the cache counters break out a subset of it. The full prompt
  is just ``input``.

:class:`Usage` is defined as disjoint, so an inclusive provider's counts must have their cache
tokens subtracted out. Routing every provider through :func:`normalize_usage` keeps the
invariant in one place instead of re-deriving it in each adapter.
"""

import logging
from collections.abc import Mapping
from typing import Any

from ..types._usage import as_token_count
from ..types.event_loop import Usage

logger = logging.getLogger(__name__)


def token_detail(details: Any, field: str) -> int:
    """Read a per-type token count out of a provider's token-details payload.

    Vendor SDKs model these breakdowns (``prompt_tokens_details``,
    ``completion_tokens_details``, and their Responses API equivalents) as objects, while
    OpenAI-compatible gateways that hand back raw JSON leave them as mappings. Both are read so a
    gateway's counts are not silently dropped.

    Args:
        details: The token-details payload, or ``None`` when the provider omits it.
        field: The wire name of the count to read, e.g. ``"cached_tokens"``.

    Returns:
        The count, or ``0`` when it is absent or not numeric.
    """
    if isinstance(details, Mapping):
        return as_token_count(details.get(field))
    return as_token_count(getattr(details, field, None))


def cache_read_count(usage: Any, prompt_details: Any) -> int:
    """Read a cache hit off an OpenAI-compatible usage payload, whichever name carries it.

    A gateway fronting Anthropic lifts the two cache counters onto the usage object as a pair, and
    builds the details payload only from counts that arrive as ints — so one reporting a count as a
    JSON float leaves the details absent while both counters sit on the usage object. A
    DeepSeek-shaped gateway spells the same count ``prompt_cache_hit_tokens``.

    Reading a single name leaves a hit folded inside the prompt count, billed at the full input rate
    rather than the cache-read rate, and nothing downstream can see it: the counters still sum to the
    reported total, so the disjointness check does not fire.

    Args:
        usage: The vendor usage payload, which may carry the count as an attribute.
        prompt_details: The payload's prompt token-details breakdown, or ``None``.

    Returns:
        The number of cache read tokens, or ``0`` when the payload reports none.
    """
    return (
        token_detail(prompt_details, "cached_tokens")
        or as_token_count(getattr(usage, "cache_read_input_tokens", None))
        or as_token_count(getattr(usage, "prompt_cache_hit_tokens", None))
    )


def cache_write_count(usage: Any, prompt_details: Any) -> int:
    """Read a cache write off an OpenAI-compatible usage payload, whichever name carries it.

    The same count arrives under three names. A gateway fronting Anthropic lifts it to the usage
    object as ``cache_creation_input_tokens``, while one speaking the OpenAI shape leaves it on the
    prompt details under either its own ``cache_creation_tokens`` or OpenAI's ``cache_write_tokens``.

    More than one can carry a value at once, so the order is the contract rather than a convenience:
    the count on the usage object is read first because it is the one the gateway itself derived,
    and a details payload can carry a differently-derived number under the same name.

    Reading a single name leaves a write folded inside the prompt count, billed at the full input
    rate rather than the cache-write rate, and nothing downstream can see it: the counters still sum
    to the reported total, so the disjointness check does not fire.

    Args:
        usage: The vendor usage payload, which may carry the count as an attribute.
        prompt_details: The payload's prompt token-details breakdown, or ``None``.

    Returns:
        The number of cache write tokens, or ``0`` when the payload reports none.
    """
    return (
        as_token_count(getattr(usage, "cache_creation_input_tokens", None))
        or token_detail(prompt_details, "cache_creation_tokens")
        or token_detail(prompt_details, "cache_write_tokens")
    )


def normalize_usage(
    *,
    input_tokens: Any,
    output_tokens: Any,
    cache_read_tokens: Any = 0,
    cache_write_tokens: Any = 0,
    reasoning_tokens: Any = 0,
    input_includes_cache: bool,
) -> Usage:
    """Normalize provider token counts into a disjoint :class:`Usage`.

    Args:
        input_tokens: The provider's prompt token count.
        output_tokens: The provider's completion token count, including reasoning tokens.
        cache_read_tokens: Prompt tokens served from the cache.
        cache_write_tokens: Prompt tokens written to the cache.
        reasoning_tokens: Tokens spent on internal reasoning. A subset of ``output_tokens``,
            reported for visibility and never added to the total.
        input_includes_cache: Whether ``input_tokens`` already contains the cache counts. Pass
            ``True`` for providers reporting the full prompt in their input field (OpenAI,
            Google, LiteLLM), ``False`` for providers reporting only net new tokens (Bedrock
            Converse, Anthropic).

    Returns:
        A :class:`Usage` whose four billed counters are disjoint and sum to ``totalTokens``.
        The cache and reasoning keys are present only when non-zero, so a provider that does
        not report them produces the same shape it always has.

    Note:
        Counts are coerced, so a missing or non-numeric value from the provider is treated as zero
        rather than raising.
    """
    input_tokens = as_token_count(input_tokens)
    output_tokens = as_token_count(output_tokens)
    cache_read_tokens = as_token_count(cache_read_tokens)
    cache_write_tokens = as_token_count(cache_write_tokens)
    reasoning_tokens = as_token_count(reasoning_tokens)

    net_input_tokens = input_tokens
    if input_includes_cache:
        net_input_tokens -= cache_read_tokens + cache_write_tokens
        if net_input_tokens < 0:
            # A provider reporting cache counts exceeding its own prompt total is
            # self-inconsistent; clamp so the invariant holds rather than emit a negative.
            logger.warning(
                "input_tokens=<%d>, cache_read_tokens=<%d>, cache_write_tokens=<%d> | "
                "cache tokens exceed reported input tokens | clamping net input to zero",
                input_tokens,
                cache_read_tokens,
                cache_write_tokens,
            )
            net_input_tokens = 0

    usage: Usage = {
        "inputTokens": net_input_tokens,
        "outputTokens": output_tokens,
        "totalTokens": net_input_tokens + output_tokens + cache_read_tokens + cache_write_tokens,
    }
    if cache_read_tokens:
        usage["cacheReadInputTokens"] = cache_read_tokens
    if cache_write_tokens:
        usage["cacheWriteInputTokens"] = cache_write_tokens
    if reasoning_tokens:
        usage["reasoningOutputTokens"] = reasoning_tokens

    return usage
